### ISSUE TITLE
Redefine Release steps during release timeline

### ISSUE THREAD

**Issue by Ressetkk** [kyma-project/test-infra#4971]
We had a pretty intense internal brainstorming session about new release process. This issue is an attempt to break down parts of Epic to more specific tasks. This issue tackles the problem of the current release process that release master has to do during the Release timeline.

# Current process
Right now we use the following process for release. This is roughly based on current release documentation available at https://kyma-project.io/community/guidelines/releases-guidelines/02-kyma-release-process/

## Overview
Once there is a decision about releasing a new version, release master cuts off a branch by issuing the following commands against main branch:
```
git checkout release-x.y
git push -u upstream release-x.y
```
Once the the branch is pushed they create a PR that bumps the VERSION file on the root of repository. After gathering required approvals and merging the PR one creates the release by going into GitHub Actions page and issuing manual workflow called `Release Kyma` against release-x.y branch. If the workflow is approved by the manager the Release is being made. 
**For Kyma 1.x release additionally file resources/installer/installer.yaml needs to be bumped along with VERSION file!**

During testing phase Release master actively bumps VERSION file and runs new workflows `Release Kyma` to actively release the following RCs and ultimately actual release.
In addition doing release creates test cluster that release master needs to fetch credentials from and provide it to testing teams on Slack.

That concludes steps required by release master to perform a release.

## Problems with above flow
* We use 2 CI infra - Prow and Github Actions. Only one CI infra should be used - Prow preferably. GitHub Action is used only to block people from doing unintended releases by pushing a tag to the repository. Once the Write access to the repository will be revoked we will be able to migrate to something else, or define small group of people who are responsible for releases.
* We use VERSION file to define current version. If not bumped correctly this makes some of the jobs fail, like kyma-artifacts.

# Proposals

## Terminology
* Release manager - decisive person that makes decisions about Release
* Release master - person that actually makes a release and watches pipelines

Below there are improvement ideas that we came up with. I tried to group and define them as good as possible. We aim to make the release automatic and as simple as using only several commands.

## The "easiest" way.
This is proposal made by me. It operates on current infrastructure and tooling, but implies we have write access revoked for org members AND VERSION file bumping is already removed #4967. Only defined people have write access to the repository, eg. Administrators and release master. The idea is to operate on release using Tagging with command `git tag`, or using GitHub UI `Create Tag`.

Once there is a decision about releasing a new version, release master cuts off a branch by issuing the following commands against main branch:
```
git checkout release-x.y
git push -u upstream release-x.y
```
Once the branch is cut-off the release master creates a *release candidate* tag based on latest commit from cut-off branch.
```
git tag -a x.y.z-rc1 -m "Release x.y.z-rc1" # message can be more complex. subject of discussion
git push -u upstream x.y.z-rc1
```
After pushing release candidate tag execute set of automated workloads - postsubmit ProwJobs.
* Create pre-release on GitHub
* Create test cluster and notify testing teams on Slack with credentials to test cluster
* Promote release artifacts
* Update CHANGELOG.md on main branch.
* Any other desired workflow

Now the testing phase starts - Developers test product and raise cherry-picks from main to the release branch with required fixes. However, release lead needs to approve every PR that lands into `release-x.y` branch. Notification is automated, and PRs are labelled by Prow. 
Once the desired set of PRs is approved and the decision about next release is made, release master creates a tag again.
```
git tag -a x.y.z-rc1 -m "Release x.y.z-rc2"
git push -u upstream x.y.z-rc2
``` 
The testing process is repeated. and RC number is incremented accordingly.

After fixing all bugs and issues final release is made. Release master creates final tag:
```
git tag -a x.y.z-rc1 -m "Release x.y.z"
git push -u upstream x.y.z
```
This runs the following automated actions:
* Create release on GitHub
* Create test cluster and notify testing teams on Slack with credentials to test cluster
* Promote release artifacts
* Update CHANGELOG.md on main branch.
* Any other desired workflow

If there would any need for patch release, be it emergency or regular one, Devs raise PRs with cherry-picks which are then approved by release manager. Release master then makes a tag with patch version - the same way as regular release.

This flow works pretty well for kyma-project/cli. It would be good to have unified process across all org repositories.
**This gives full control over which commit is considered a tag. Release manager should be able to rebase commits to change the history and rewrite it on release branch if possible. Then create a tag for specific commit SHA.**
## The "issue" way
This proposal runs around creating Issue that has control over entire release flow. Release master, or person that belongs to `release-masters` team issues commands that operate on Prow plugin that manages releases. People not in this group will not be able to use plugin commands.
**This requires to develop separate tooling that will operate on issue comments.**

Release master creates issue called `Release x.y.z`. Then using comment commands issues branch cut-off
```
/cutoff source destination
/release branch x.y.z-rc1
```
Would instruct bot to cutoff *source* branch and name it *destination*. Then after creating branch create release from that branch and name it `x.y.z-rc1`.

Making release will run the same workloads as when pushing a tag manually.

Testing phase looks the same as in the previous proposal.

Once all PRs are collected and approved by release manager, release master releases next RCs incrementing RC number.
```
/release branch x.y.z-rc2,3,4 etc.
```

If the latest RC is final. Release master creates final release issuing command
```
/release branch x.y.z
```
Once the process is done and Release is made issue is closed.

Every release needs to be issued on separate issue, be it Patch Release or a new Minor release.

# Conclusion
We need to decide on which route we should follow. I prefer the easiest way, because we don't need to develop additional tooling and we can use what we have right now, drastically reducing man hours required to implement the improvement. Also that approach is quite straightforward, because generally everyone is accustomed to use `git`, and if they were chosen as release masters there should be no problem.

Let's keep the discussion about process itself under this topic.

#4799

**Comment by Ressetkk** [comment]
I have prepared yet another approach to this topic. What can be used for the overall product release is using GitFlow with GitVersion responsible for automatic tagging the release.

## GitFlow with GitVersion
For this approach we introduce several long-living branches:
* main - branch that contains the latest released and tagged version
* develop - default branch with active development towards next version
* support/Y.x - support branches when previous major version requires support

And several short-living branches that are removed once get merged:
* release/x.y.0 - release branch for *minor* version where we fix bugs and security vulnerabilities
* hotfix/x.y.z - branch created when emergency patch release is required.
* feature/feature-name - branch for long-term features where multiple people work on (optional)

The flow depends heavily on making separate branches for different lifecycles of the version.
Pros:
* Strict to maintain
* Ability to release hotfixes at any time of the release lifecycle
* Clear and precise tagging
* Releasing a version (tagging) can be automated
* In compliance with Continous Delivery definitions
* Can be blocked by 

Cons:
* Requires manual cutoff of each branch
* Requires some knowledge around git and versioning
* Resolving conflicts may be really difficult in rare cases
* Requires more configuration around Prow for more branches

We also assume that tagging works 100% automatically and there is no human interaction, apart from creating the required branch.

## Example flows
I will try to briefly showcase the flows that can be achieved. For the example I will use current Kyma versions.
### Releasing a minor version
The diagram below shows the example flow that can be achieved with the GitFlow. We assume that main and develop start from the same HEAD.
```mermaid
sequenceDiagram
Note over main:2.0.0
develop->>develop: commits
Note over develop: 2.1.0-alpha.1
develop->>develop:commits
Note over develop: 2.1.0-alpha.2
develop-->>release/2.1.0:release cut-off
develop->>develop:commits
Note over develop:2.2.0-alpha.1
Note over release/2.1.0: 2.1.0-beta.1
release/2.1.0->>release/2.1.0:commits
Note over release/2.1.0: 2.1.0-beta.2
release/2.1.0-->>main:merge
Note over main:2.1.0
release/2.1.0-->>develop:merge
Note over develop:2.2.0-alpha.2
Note over release/2.1.0:The branch gets removed
```

### Releasing a hotfix for current version
If the major version needs emergency patching the flow should be possible with the approach above.
```mermaid
sequenceDiagram
main-->>hotfix/2.1.1:hotfix branch from main
hotfix/2.1.1->>hotfix/2.1.1:commits
Note over hotfix/2.1.1:2.1.1-beta.1
hotfix/2.1.1->>hotfix/2.1.1:commits
Note over hotfix/2.1.1:2.1.1-beta.2
hotfix/2.1.1-->>main:merge
Note over main:2.1.1
hotfix/2.1.1-->>develop:merge
Note over hotfix/2.1.1:hotfix branch is deleted
```

### Releasing major version
Releasing major version would work similarly like with minor version.
```mermaid
sequenceDiagram
Note over main:1.24.11
develop-->>release/2.0.0:release cut-off
develop->>develop:commits
Note over develop:2.1.0-alpha.1
release/2.0.0->>release/2.0.0:commits
Note over release/2.0.0:2.0.0-beta.1
release/2.0.0->>release/2.0.0:commits
Note over release/2.0.0:2.0.0-beta.2
release/2.0.0-->>main:merge
Note over main:2.0.0
release/2.0.0-->>develop:merge
Note over develop:2.1.0-alpha.2
Note over release/2.0.0:release branch is deleted
```

## Support branches
Support branches try to resolve the problem, where the main branch is already on next major release (eg. 2.0.0), but the version 1.x still needs patching. The approach is to create a separate branch called `support/1.x` that is created from the latest tag of 1.x release.

### Minor release
Let's say we need to release 1.25 version of Kyma and we have released 1.24.11 release and 2.0.0 as latest. The flow would look as follows:
```mermaid
sequenceDiagram
Note over main:2.0.0
main-->>support/1.x:branch from tag 1.24.11
support/1.x-->>release/1.25.0:release cut-off
release/1.25.0->>release/1.25.0:commits
Note over release/1.25.0:1.25.0-beta.1
release/1.25.0->>release/1.25.0:commits
Note over release/1.25.0:1.25.0-beta.2
release/1.25.0-->>support/1.x:merge
Note over support/1.x:1.25.0
Note over release/1.25.0:the branch is deleted
```

### Hotfix release
```mermaid
sequenceDiagram
Note over main:2.0.0
main-->>support/1.x:branch from tag 1.24.11
support/1.x-->>hotfix/1.24.1:release cut-off
hotfix/1.24.1->>hotfix/1.24.1:commits
Note over hotfix/1.24.1:1.24.1-beta.1
hotfix/1.24.1-->>support/1.x:merge
Note over support/1.x:1.24.1
Note over hotfix/1.24.1:the branch is deleted
```

## Feature branches
With widely adopted forks across all contributors in Kyma there would be no real need of introducing feature branches on upstream repo. Every contribution or PR can be raised for develop, or current working support/  branch.
```mermaid
sequenceDiagram
Note over develop:2.1.0-alpha.1
develop-->>pr/feature:Create fork branch
pr/feature->>pr/feature:commits
pr/feature-->>develop:Open Pull Request
loop Review process
develop-->>pr/feature:PR has requested changes
pr/feature->>pr/feature:commits
end
pr/feature-->>develop:merge accepted PR
note over develop:2.1.0-alpha.2
note over pr/feature:Fork branch is closed
```

If there was a need to create a feature branch on upstream repo for bigger topic, then the flow would look the same as above. 

## Conclusion
With GitFlow approach it should be possible to have full control over the version lifecycle at basically any time of the release flow. However it would require quite the effort to maintain and resolve any conflicts that would emerge during merge processes. This would definitely enable us for introducing some nice automations, allow setting-up continuous delivery and promotion pipelines with automatic versioning of each release.

**Comment by dekiel** [comment]
Recently we discuss Kamil's proposal and agreed that we would like to stick to our current process and try to change it slightly to address our requirements. Because of that I've prepared following two proposal which are based on our current flow.

Proposal 1.

The flow for major/minor and patch releases is same as current one. It extend our flow with strategy for an emergency release. A emergency release will be prepared and released from separate branch. This branch will be always cut off from latest release tag on a release branch and merged back to release branch after emergency release is done.

pros:
- We keep our current flow and only extend it.
- In case emergency release during patch release, we don't have to cancel current patch release and move commits to next, incremented patch release.
- Less conditions to evaluate and less actions to do for emergency release so it should be easier to automate.
cons:
- Emergency release is little different than patch release.

![release_proposal_2](https://user-images.githubusercontent.com/12400578/158217715-ad77cf13-b601-4331-8259-c5ef886de460.png)

Proposal 2.

The flow for major/minor release remain unchanged. A strategy for patch and emergency release is the same. In both cases we will cut off new branch from latest release tag on a release branch. Patch/emergency release branch i merged back to release branch just before release. Release is always done from release branch. In case an emergency patch is done during on going patch release, a patch release branch i closed and commits are moved to next patch release immediately after emergency release is done.

pros:
- A patch and emeregency release use the same flow.
cons:
- In case emergency release during patch release we need to cancel existing patch release and move it to the new one. This require additional steps.

![release](https://user-images.githubusercontent.com/12400578/158242606-28bd0ded-671e-41c9-a97c-1d7459dea548.png)


In terms of increasing release process automation and addressing release 1.24 post mortem action items we would like to work on implementing following process.

We made few main assumptions.

1. Release issue is an interface to interact with release and get status updates about release progress.
2. Component and release prowjobs are defined as inrepo config. That means their definitions are stored in a repositories for which a release process is running and are versioned along with a code. This allow us to drop step for generating prowjob definitions for each supported release in test-infra main branch.
3. Scripts are still maintained under test-infra repository.
4. VERSION file will be removed under #4967 . We can drop steps about updating content of this file.
5. Post mortem action item 2 will be implemented by changes in branch strategy.
6. Post mortem action item 9 will be implemented by changes in release automation.

Proposed release process.

1. Release master create Release Issue from template.
   - We distinguish three types of releases, major/minor, patch and emergency/hotfix release.
   - Release Issue will cary all configuration and human input needed for the release process.
   - It will represent a current status of release process and collect references to all other issues, PRs and artifacts related to the release process.
   - Issue will be added to the Github Projects of all Kyma teams.
   - Issue can be added to a new Github Project named Kyma Release which can contain only issues related to the Kyma release process to represent their states in one place.
   -  Creation of this issue will trigger a release process.
   -  Only members of releaseMaster owners alias group in a repo which is released are allowed to create such issue and trigger release. Issues opened by other users will be closed with relevant comment.
2. Create a release branch in the test-infra repository. We still need version test scripts.
3. [major/minor only] Create a release branch in the kyma repository. The name of this branch should follow the release-{major}.{minor} pattern, such as release-2.4 if it not exist yet.
   - Creation of branch is hold until all issues for present PR images are closed.
   - Update release issue with status and provide copy paste command to checkout branch on developers workstation.
   - Send notification to the release slack channel
4. [major/minor only] Create PR to generate prowjobs for release. This will update inrepo prowjobs test-infra extra refs with release branch name.
   - Run rendertemplates in released repository release branch
   -  On a PR all jobs syntax tests should run. Which jobs to run? test-infra/development/jobs/make jobs-tests
   -  Check if the main branch contains any PR-images. If PR images are found:
       - automated notification should be send to responsible teams,
       -  github issue for correction should be raised, assigned to the team and release project and linked to the release issue.
5. [emergency only] If another patch release for the same version is in progress. This is needed in case we start an emergency/hotfix release.
    - Close and cleanup objects related to canceled release.
7. [emergency only] Create a patch or emergency/hotfix release branch in the kyma repository.
    - The name of this branch should follow the release-{major}.{minor}.{patch} pattern, such as release-2.4.1. For emergency release we could add suffix "-hotfix"
    - Branch is created from latest release tag.
    - Update release issue with status and provide copy paste command to checkout branch on developers workstation.
    - Send notification to the release slack channel
8. [start release candidate cycle ] Create release candidate tag.
   - Following tests were running on VERSION file change PR. What else tests we need to run here, see list of all prowjobs for release-2.0 branch in notes?
     - pre-rel20-kyma-artifacts
     - pre-rel20-kyma-gke-integration
     - pre-rel20-kyma-gke-upgrade
     - pre-rel20-kyma-integration
     - pre-release-pr-image-guard (this should run always to protect branch from PR images.)
   - Update release issue about result of tests.
   - If any automated test fail, create issue for fix and link it to the release issue.
9. Create release candidate cluster.
   - update release issue and send notification to release slack channel
10. Create Github Issues for manual tests.
   - add issues to respective teams projects and release project
   - link issues to the release issue.
   - manual test issues Can be auto assigned to POs of respective teams.
   - if manual test failed, developer create issue for bug fix.
11. [Start of bug fix development process towards the release - this cycle remain the same as now] Create a bugfix-branch based on the release branch you want to extend. Add your changes and create a Pull Request.
12. Once you create a Pull Request to the release branch, the set of checks is triggered. These jobs run in the same way as jobs that run on every Pull Request to the main branch. If you create a Pull Request that contains changes to the components, the component-building job is triggered. If you make any changes in the charts, the integration tests are triggered.
13. If you detect any problems with your PR, fix the issues until your checks pass.
14. After all checks pass, merge your PR to the release branch. Merging the PR triggers the post-submit integration tests automatically.
15. [End of bug fix development process towards the release] If there's a need for additional changes in the release branch during the development process, open a new PR to the release branch. Repeat steps 10-14 for this PR.
16. [End of Release candidate cycle] Release master issue a command on Release issue to create tag and RC cluster. /createReleaseCandidate.
    - close manual test issues for previous release candidate.
    - start release candidate cycle again
    - This can be automated and create next release candidate on condition if all issues for bug fixes were closed. This requires all development against release branch is represented by at least one github issue.
17. Release master issue command on release Issue to create a release, /createRelease.
    - command can be issued only by members of release master aliases group
    - Depending on chosen branching scenario, some additional steps might be executed here.
18. All remaining steps are the same as in current process starting from creating release. Following are some changes we will introduce.
    - all steps will run on Prow not as Github actions,
    - no waiting for approvals, as it was done in step 16,
    - in case post release upgrade job failed, creating a github issue,
    - notifications will send where relevant
    - before notifying about CLI release a cleanup activity will take place.

**Comment by pbochynski** [comment]
@dekiel I am slightly in favor of proposal 2 (patch and emergency releases use the same process) because this way we are sure that emergency release flow actually works. I would like to avoid the situation when we are in emergency mode and we discover some issues in the hotfix process.

**Comment by Sawthis** [comment]
@dekiel @pbochynski 
I like more Proposal 1 as it is easier to understand and maintain (it has less additional steps). We can have the same process for patch and hotfix release just from different branch.

**Comment by dekiel** [comment]
I prefer a proposal 1. Because it doesn't require closing ongoing patch release branch, opening new one just after the hotfix is done and moving commits between old and new patch release branches. All this steps are needed in case hotfix is required when we have ongoing patch release. From developer point of view, proposal 1 let developers bring commits in to release branch at any time not only when a release is started. In proposal 2, developer must wait for start of release and patch release branch cutoff.

However having a process and tools which are not tested and confirmed they work, especially for emergency release is bad idea. I think we can address this by exercising an emergency release flow for normal patch release. Such exercise could be done every 6 months. This way we could test whole process and tools exactly how we will use them during real emergency release. @pbochynski @Sawthis what do you think?

**Comment by Sawthis** [comment]
@dekiel I like the idea but I think the 6 months intervals are too long.

**Comment by pbochynski** [comment]
> However having a process and tools which are not tested and confirmed they work, especially for emergency release is bad idea. I think we can address this by exercising an emergency release flow for normal patch release. Such exercise could be done every 6 months. This way we could test whole process and tools exactly how we will use them during real emergency release. @pbochynski @Sawthis what do you think?

I think you can keep all the pros from proposal 1 and have one tooling for it. What about you ALWAYS create an additional hotfix/patch branch that will be merged back to release branch. If it is regular patch release you can create it directly from the release branch head. If it is emergency branch you create it from latest tag in that branch. 

The flow would look like this:

Patch release | Hot fix release
---|---
| 1. create branch  from latest commit in release branch | 1. create branch from latest tag in release branch |
| **2. do  nothing**  | **2. cherry pick / commit emergency fixes** |
| 3. do the release | 3. do the release
| 4. merge (nothing to merge) | 4. merge (fix lands in the release branch)

**Comment by Sawthis** [comment]
@pbochynski @dekiel According to the latest strategy about backporting security vulnerabilities announced on [Slack](https://sap-btp.slack.com/archives/C01KKPXCPK8/p1647937124006879?thread_ts=1647596934.960849&cid=C01KKPXCPK8) we will have patch releases only for critical issues. So in this case I think we can release hotfix just from release branch.

**Comment by dekiel** [comment]
Implementation of new branching strategy for releases will be done under #5407
